### DISCUSSION TITLE
What about HarmonyOS?

### DISCUSSION THREAD

**Discussion by inson1** [AvaloniaUI/Avalonia#14019] [General]
Will there be support?

**Comment by thevortexcloud** [comment]
It sounds like it supports android apps? In which case it probably already works.

**Reply by inson1** [reply]
@thevortexcloud no it cant  https://www.gizchina.com/2023/11/21/how-huawei-plans-to-keep-android-app-support-for-global-users/

**Reply by thevortexcloud** [reply]
It can right now. The version that drops android support is not even getting a dev preview until next year.

In any case, unless it supports .NET and Skia there is not much anybody can do to get it running on the future release.

**Reply by kyuranger** [reply]
I'm sure it supports Skia because it supports Flutter (but support for HarmonyOS is maintained by huawei not Google). But I haven't heard of relevant support for .Net, but Unity does support it, but I'm not sure whether Unity's .Net and the .Net we often talk about are the same thing.

**Reply by thevortexcloud** [reply]
Unity uses a heavily modified version of Mono. It's not really meant to be used as a general runtime. Unity also has a .NET transpiler that converts .NET code to [C++](https://docs.unity3d.com/Manual/IL2CPP.html)/native code. Although they are trying to [move away from Mono.](https://blog.unity.com/engine-platform/porting-unity-to-coreclr)


In any case, Avalonia still needs a compatible .NET runtime. If someone ports Mono to it (or it already works?), Avalonia will probably also work to some extent.

**Comment by kyuranger** [comment]
I remember reading an official Avalonia article before, saying that it is easier for Avalonia to be compatible with new platforms than other frameworks. As long as that framework supports image display, it can easily connect to Avalonia's display interface. So I think it's not that Avalonia can't do it (just like when Vison OS just came out, Avalonia quickly supported it), but whether the Avalonia team has the energy to be compatible with one more platform.

As far as I know, HarmonyOS is currently only supported by Flutter (not officially supported by Google), Unity, and Huawei's official framework. If Avalonia supports it, that would be a good thing, but I personally think that based on various issues such as HarmonyOS market share, it is unlikely to attract the Avalonia team to support it.

**Reply by thevortexcloud** [reply]
> As long as that framework supports image display, 

It's more complicated than that. The platform needs to allow for directly drawing pixels to the screen, as Avalonia is a drawn control framework. Typically this is done via Skia. But it's possible to wire up a different backend. The platform also needs to support .NET/be compatible with .NET AOT, ideally Skia, and support some sort of user input (EG touch/mouse and keyboard). Vision OS specifically is directly compatible with iOS apps. So it's not much of a leap to get apps working on it as Avalonia already supports iOS (as does .NET).

In any case, HarmonyOS can _currently_ run android apps. So Avalonia should already work on it.

**Reply by kyuranger** [reply]
Compatibility with Android is just a way for Huawei to gain market share for HarmonyOS. As its market share gradually increases, it will gradually develop a system based on Android, but it will be very different from Android. It's somewhat similar to the relationship between Unix and iOS. So I think Avalonia supports Android, but as HarmonyOS begins not to support Android applications, it does not mean that Avalonia can support HarmonyOS. Of course, it is meaningless how we talk about this. What is important is what the Avalonia team thinks (even though I don't think the Avalonia team will give any reply on this topic).

**Comment by dongfo** [comment]
Even if the Avalonia view rendering can run on the HarmonyOS, it would be meaningless if the dotnet itself cannot call the HarmonyOS Api

**Reply by kyuranger** [reply]
It seems that HarmonyOS is developed based on Android, so I think support should not be a big problem.

**Reply by 996666925** [reply]
No, there is no relationship between HarmonyOS and Android

**Comment by bxjg1987** [comment]
Now (November 2024) HarmonyOS is a completely independent system, unrelated to Android. This system has received strong support from the Chinese government, and its market prospects in mainland China are very good. It would be great if Avalonia could support it.
It also has an open-source version of OpenHarmonyOS.

**Reply by MikeCodesDotNET** [reply]
For this to be considered, we would need to know the following:

- Does this new OS support .NET? 
- Does it support Skia? 
- How much demand is there? 
- Who is going to fund this work?

**Reply by CeSun** [reply]
Hi! I'm trying to port Avalonia to HarmonyOS, so I can answer these questions.

> Does this new OS support .NET?

Yes, HarmonyOS supports linux-musl dynamic libraries. I have successfully run the dynamic libraries pulished by NativeAot on HarmonyOS.

> Does it support Skia?

I don't know, when porting Avalonia, no exception occurs when calling the UseSkia function

> How much demand is there?

In the future, perhaps all Huawei devices will use this system, currently, only a few Huawei flagship devices can be upgraded to HarmonyOS.

> Who is going to fund this work?

I'm sorry, I can only answer that there is no one

I am an amateur .net developer. I am trying to port the Avalonia project to HarmonyOS. Here are the related repositories:

[https://github.com/CeSun/OpenHarmony.Avalonia](https://github.com/CeSun/OpenHarmony.Avalonia)
[https://github.com/CeSun/OpenHarmony.Net](https://github.com/CeSun/OpenHarmony.Net)

The principle is: I called the functions related to HarmonyOS's canvas (it is called XComponent on the HarmonyOS) into C#, and used P/Invoke to import some native APIs of the HarmonyOS. Currently, the rendering loop function and EGL-related functions can be obtained in C#.

But Avalonia has not been successfully drawn on the system, and the feedback to me is just a white screen.

If it is convenient, please give me some guidance. I am happy to participate in open source projects.

Other questions you may want to know:
1. What is OpenHarmony?
   An open source version of HarmonyOS. It has the same API as HarmonyOS, but the system interface looks very simple.
2. Does coreclr support running on HarmonyOS?
   Theoretically, it is possible, but I am trying it for now.

**Reply by kekekeks** [reply]
>But Avalonia has not been successfully drawn on the system, and the feedback to me is just a white screen.

From a brief look, you are missing a dispatcher implementation that essentially defines how avalonia interacts with the os-specific event loop or drives its own one.

Take a look at various IDispatcherImpl implementations in our repo and see which looks closer to what harmony event loop can provide.

**Reply by thevortexcloud** [reply]
Also https://github.com/MrJul/Estragonia is probably useful of an example of how to implement a standalone Avalonia backend.

**Reply by kekekeks** [reply]
For render timer you probably need to reuse UiThreadRenderTimer initially, since multithreaded rendering is usually a tricky part to get right

**Reply by CeSun** [reply]
@kekekeks @thevortexcloud Many thanks!

**Reply by kekekeks** [reply]
Actually, since harmony seems to expect you to render in a per-view frame callback, it's probably better to have a separate compositor per view, since you can't share a single render timer between those.

But I think that for the most part you are missing a dispatcher implementation: composition commits are might be not actually sent to the compositor.

**Reply by kekekeks** [reply]
BTW, I assume that debugging story is nonexistant? I. e. you are forced to build the app into a NativeAOT library and load is as a native component for a harmony app

**Reply by CeSun** [reply]
> I assume that debugging story is nonexistant?

Yes, I compiled the avalonia application into a native library, and then called this library in the Harmony project.
For debugging, in theory, Harmony uses the lldb debugger. If a crash occurs, I can see the stack symbols on the Harmony IDE, but the Harmony IDE cannot set breakpoints to the C# source code. Maybe I can explore using vscode debugging later.

**Reply by thevortexcloud** [reply]
Out of curiosity, is there any reason you can't just run the actual .NET runtime? Hosting it inside another app is fully supported.

https://learn.microsoft.com/en-us/dotnet/core/tutorials/netcore-hosting

It might even give you access to proper debugging too. Although I do consider this approach to be a hack at best.

It's not something I have tried myself though.

**Reply by CeSun** [reply]
> Out of curiosity, is there any reason you can't just run the actual .NET runtime? Hosting it inside another app is fully supported.
> 
> https://learn.microsoft.com/en-us/dotnet/core/tutorials/netcore-hosting
> 
> It might even give you access to proper debugging too. Although I do consider this approach to be a hack at best.
> 
> It's not something I have tried myself though.

It seems that the hostfxr-related APIs are designed for PC. I don't know which folder I should put the BCL dlls in so that coreclr can load them correctly.
So in order to quickly verify the feasibility of Avalonia, I chose to use the NativeAOT runtime, which is the fastest solution.

**Reply by thevortexcloud** [reply]
Fair enough. I don't know enough hosting the runtime or HarmonyOS to help with that. I imagine it's possible though if it's able to just run standard Linux Musl binaries.

**Reply by kekekeks** [reply]
>It seems that the hostfxr-related APIs are designed for PC. I don't know which folder I should put the BCL dlls in so that coreclr can load them correctly.

Just use a _self-contained_ (not single file) published app, it will have required libs including libhostfxr.so that would correctly resolve coreclr location.

You can find a simple example here, it doesn't require a shared CLR runtime installed: https://github.com/kekekeks/simple-clr-host/blob/master/main.cpp

Note that that particular host is designed to be a _replacement_ for the default one, so adjust `GetExePath` accordingly.

**Reply by kekekeks** [reply]
i. e. compile the host and like this with `TEST_EXE_PATH=`pwd`/MyApp ~/Projects/SimpleClrHost/cmake-build-debug/SimpleClrHost`.

**Reply by kekekeks** [reply]
You can call `hostfxr_run_app` from a background thread and let the C# code to P/Invoke into your host to set up callback pointers and then `Thread.Sleep(-1)`

**Reply by kekekeks** [reply]
Given that harmony uses Node internally, you can try to get this to work too - https://github.com/sanosdole/nodeclrhost

**Reply by CeSun** [reply]
> For render timer you probably need to reuse UiThreadRenderTimer initially, since multithreaded rendering is usually a tricky part to get right

Some controls are rendered successfully (TextBlock, Image, Rectangle, etc.), but controls like TextBox, Button, etc. will cause the sk_canvas_flush call to crash.

![QQ_1731031169013](https://github.com/user-attachments/assets/e7d02a0b-0d5e-4bd9-a594-2e0c5759358e)

**Reply by thevortexcloud** [reply]
I am not an expert on Skia/Avalonia, but this looks like some GPU problem? I could be completely wrong though. But if it is,  it's entirely possible the problem is outside your code. A lot of graphics drivers, especially on mobile devices are supposedly extremely broken/barely functional. You could try some sort of software renderer, either in Avalonia or in the OS (EG forcing a software rendering graphics library).

**Reply by CeSun** [reply]
> I am not an expert on Skia/Avalonia, but this looks like some GPU problem? I could be completely wrong though. But if it is, it's entirely possible the problem is outside your code. A lot of graphics drivers, especially on mobile devices are supposedly extremely broken/barely functional. You could try some sort of software renderer, either in Avalonia or in the OS (EG forcing a software rendering graphics library).

Great! I switched to software rendering and it worked fine! But I still have to investigate why skiasharp crashes when using opengl as the backend.

Many thanks

**Reply by thevortexcloud** [reply]
Glad I could help. Avalonia does also have a Vulkan renderer which might be of use.

https://github.com/AvaloniaUI/Avalonia/pull/12737

> But I still have to investigate why skiasharp crashes when using opengl as the backend.

My guess is either a SkiaSharp bug or bad GPU drivers. OpenGL on mobile is also typically a subset of the full OpenGL spec. 

https://en.wikipedia.org/wiki/OpenGL_ES

It's entirely possible Skia is expecting the full OpenGL APIs to be available (since it thinks its on Linux) and HarmonyOS has not implemented it. But your guess is as good as mine as to what the real problem is.

**Reply by kekekeks** [reply]
Could be context not being current when Skia tries to access it (note that this can also happen from a SKSurface finalizer)

**Reply by CeSun** [reply]
> Could be context not being current when Skia tries to access it (note that this can also happen from a SKSurface finalizer)

Final conclusion: I used the linux-arm64 skiasharp library incorrectly, The linux-musl-arm64 library should be used on the Hongmeng platform😅

**Reply by thevortexcloud** [reply]
I have made a similar mistake before, so much time wasted... But glad you figured it out!

**Comment by Coloryr** [comment]
First of all, are you asking about pure `HarmonyOS NEXT` or `HarmonyOS`, the `HarmonyOS NEXT` use `ArkTS`, instead of java, so it needs to be supported by `dotnet-runtime` first, next is support for `skia` and GPU drivers.

**Comment by CeSun** [comment]
Project: https://github.com/CeSun/OpenHarmony.Avalonia
As of 2025/02/16, Avalonia's support for Hongmeng platform is as follows:

https://github.com/user-attachments/assets/304b32b6-31ad-4ea3-bdef-ba40997e35ef

**Comment by RoderickIveans** [comment]
any update?

**Reply by MikeCodesDotNET** [reply]
There’s no update to share at this time, and I don’t anticipate any changes to our supported platforms.
